File: frontend/lib/models/sale.dart
import 'enums.dart';
import 'gated_entity.dart';
import 'installment.dart';
import 'sale_payment.dart';

/// A vehicle sale to a customer — either full cash or advance + monthly
/// installments. Gated like every other entity.
class Sale with GatedEntity {
  Sale({
    required this.id,
    required this.vehicleId,
    required this.customerId,
    required this.mode,
    this.salePrice,
    this.advance = 0,
    this.monthly = 0,
    this.dueDate,
    this.saleDate,
    this.vehicleAmount = 0,
    this.additionalFitting = 0,
    this.dlCharges = 0,
    this.documentCharges = 0,
    this.otherExpenses = 0,
    this.hpAmount,
    this.remainingAmount = 0,
    this.customerWhatsapp = '',
    this.financerId,
    this.financerName,
    this.invoiceNo,
    this.closedAt,
    List<Installment>? installments,
    List<SalePayment>? payments,
    this.saleStatus = 'active',
    this.invoicePath,
    this.remarks,
    required this.createdBy,
    required this.createdAt,
    this.status = EntityStatus.active,
    this.confirmedBy,
    this.confirmedAt,
    this.rejectionReason,
    this.unsellReason,
    this.unsellStage,
    this.unsellRequestedBy,
    this.seizedAt,
    this.seizedBy,
    this.seizeReason,
    this.seizeStage,
    this.seizeConfirmedAt,
    this.seizeConfirmRemarks,
    this.seizeCancelRemarks,
    this.sold = false,
    this.pendingEdit,
    this.editStage,
    this.editRequestedBy,
  })  : installments = installments ?? [],
        payments = payments ?? [];

  @override
  final String id;
  final String vehicleId;
  final String customerId;
  PaymentMode mode;
  int? salePrice;

  // Advance + installments terms:
  int advance;
  int monthly; // per-month installment amount
  DateTime? dueDate; // first installment due date

  // Price breakdown (total = salePrice = sum of the five below).
  int vehicleAmount;
  int additionalFitting;
  int dlCharges;
  int documentCharges;
  int otherExpenses;
  // Loan case only:
  int? hpAmount; // HP / loan amount (super admin only)
  int remainingAmount; // typed by the user

  // Fields added for the real API flow:
  DateTime? saleDate;
  String customerWhatsapp;
  int? financerId; // finance company for this sale (optional)
  String? financerName; // its display name, from the API
  String? invoiceNo;
  DateTime? closedAt;

  List<Installment> installments;
  List<SalePayment> payments;

  /// The pending (awaiting-approval) payment for an installment, if any.
  SalePayment? pendingPaymentFor(String installmentId) => payments
      .where((p) => p.installmentId == installmentId && p.isPending)
      .cast<SalePayment?>()
      .firstOrNull;

  /// 'active' | 'closed' | 'cancelled' | 'rejected'
  String saleStatus;
  String? invoicePath;
  String? remarks;

  @override
  final String createdBy;
  @override
  final DateTime createdAt;
  @override
  EntityStatus status;
  @override
  String? confirmedBy;
  @override
  DateTime? confirmedAt;
  @override
  String? rejectionReason;
  String? unsellReason;

  /// Unsell approval: null | 'pending' (admin requested, awaiting super admin).
  String? unsellStage;
  String? unsellRequestedBy;

  /// An admin requested an unsell that a super admin hasn't approved yet.
  bool get isUnsellPending => unsellStage == 'pending';

  // Seizure (repossession) audit — set on a seized sale (kept as history).
  DateTime? seizedAt;
  String? seizedBy;
  String? seizeReason;

  /// Seize lifecycle: null | 'pending' (admin requested, awaiting super admin) |
  /// 'seized' (active — badge shown, cancel/confirm available) | 'confirmed'.
  String? seizeStage;
  DateTime? seizeConfirmedAt;
  String? seizeConfirmRemarks;
  String? seizeCancelRemarks;

  /// True once the user confirms a fully-paid sale as sold. While false the
  /// Seize button is available; once true it is hidden.
  bool sold;

  /// Admin requested a seize that a super admin hasn't approved yet.
  bool get isSeizePending => seizeStage == 'pending';

  /// Pending edit: an admin proposed changes that a super admin hasn't approved
  /// yet. The live sale keeps its current values; [pendingEdit] holds the
  /// proposed field values (for the approval diff).
  Map<String, dynamic>? pendingEdit;
  String? editStage;
  String? editRequestedBy;

  /// An admin requested an edit that a super admin hasn't approved yet.
  bool get isEditPending => editStage == 'pending';

  /// Seize is active — the 'Seized' badge stage (can be cancelled or confirmed).
  bool get isSeizeActive => seizeStage == 'seized';

  /// Seize has been finalised (badge cleared; vehicle is a plain free vehicle).
  bool get isSeizeConfirmed => seizeStage == 'confirmed';

  /// A sale can be reversed ("unsold") only within this window after creation.
  static const unsellWindow = Duration(days: 1);

  /// Whether the 1-day unsell window is still open. [createdAt] is a UTC
  /// instant, so compare against the current UTC time (the server and phone can
  /// be in different time zones). Mirrors the backend's own check.
  bool get canUnsell =>
      DateTime.now().toUtc().difference(createdAt.toUtc()) < unsellWindow;

  DepositType get depositType =>
      mode == PaymentMode.full ? DepositType.fullCash : DepositType.downPayment;

  int get paidCount => installments.where((i) => i.isPaid).length;
  int get totalInstallments => installments.length;

  int get collected =>
      advance + installments.where((i) => i.isPaid).fold(0, (s, i) => s + i.amount);

  int get outstanding {
    final scheduled = installments.fold(0, (s, i) => s + i.amount);
    final paid = installments.where((i) => i.isPaid).fold(0, (s, i) => s + i.amount);
    return scheduled - paid;
  }
}

File: backend/app/models/sale.py
"""Sale model — a vehicle sold to a customer (full cash or down payment)."""
from datetime import date, datetime

from sqlalchemy import (
    BigInteger,
    Date,
    ForeignKey,
    Identity,
    Integer,
    Numeric,
    SmallInteger,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base
from app.models.enums import DepositType, EntityStatus, SaleLifecycle, pg_enum


class Sale(Base):
    __tablename__ = "sales"

    id: Mapped[int] = mapped_column(BigInteger, Identity(always=True), primary_key=True)
    module_id: Mapped[int] = mapped_column(
        SmallInteger, ForeignKey("modules.id"), nullable=False
    )
    vehicle_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("vehicles.id"), nullable=False
    )
    customer_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("customers.id"), nullable=False
    )

    deposit_type: Mapped[DepositType] = mapped_column(
        pg_enum(DepositType, "deposit_type"), nullable=False
    )
    sale_price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    sale_date: Mapped[date | None] = mapped_column(Date)
    amount_received: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    remaining_amount: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    # sale-price breakdown (sale_price/total = sum of these five)
    vehicle_amount: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    additional_fitting: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    dl_charges: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    document_charges: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    other_expenses: Mapped[float] = mapped_column(
        Numeric(12, 2), nullable=False, server_default="0"
    )
    # HP (hire-purchase / loan) amount — set by super admin only
    hp_amount: Mapped[float | None] = mapped_column(Numeric(12, 2))
    monthly_amount: Mapped[float | None] = mapped_column(Numeric(12, 2))
    installment_count: Mapped[int | None] = mapped_column(Integer)
    first_due_date: Mapped[date | None] = mapped_column(Date)
    customer_whatsapp: Mapped[str | None] = mapped_column(Text)
    invoice_no: Mapped[str | None] = mapped_column(Text)
    sale_status: Mapped[SaleLifecycle] = mapped_column(
        pg_enum(SaleLifecycle, "sale_lifecycle"),
        nullable=False,
        server_default=SaleLifecycle.active.value,
    )
    closed_at: Mapped[datetime | None] = mapped_column()

    # role-gate / audit
    status: Mapped[EntityStatus] = mapped_column(
        pg_enum(EntityStatus, "entity_status"),
        nullable=False,
        server_default=EntityStatus.pending_confirmation.value,
    )
    created_by: Mapped[int] = mapped_column(BigInteger, nullable=False)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    confirmed_by: Mapped[int | None] = mapped_column(BigInteger)
    confirmed_at: Mapped[datetime | None] = mapped_column()
    rejection_reason: Mapped[str | None] = mapped_column(Text)
    unsell_reason: Mapped[str | None] = mapped_column(Text)
    # seizure (repossession): set when the vehicle is seized from this customer.
    seized_at: Mapped[datetime | None] = mapped_column()
    seized_by: Mapped[int | None] = mapped_column(BigInteger)
    seize_reason: Mapped[str | None] = mapped_column(Text)
    # seize lifecycle: NULL | 'pending' (admin requested, awaiting super admin) |
    # 'seized' (active — badge shown, cancel/confirm available) | 'confirmed'.
    seize_stage: Mapped[str | None] = mapped_column(Text)
    seize_confirmed_at: Mapped[datetime | None] = mapped_column()
    seize_confirmed_by: Mapped[int | None] = mapped_column(BigInteger)
    seize_confirm_remarks: Mapped[str | None] = mapped_column(Text)
    seize_cancel_remarks: Mapped[str | None] = mapped_column(Text)
    remarks: Mapped[str | None] = mapped_column(Text)

    installments: Mapped[list["SaleInstallment"]] = relationship(  # noqa: F821
        back_populates="sale",
        cascade="all, delete-orphan",
        order_by="SaleInstallment.month_number",
    )
    payments: Mapped[list["SalePayment"]] = relationship(  # noqa: F821
        back_populates="sale", cascade="all, delete-orphan"
    )
    # the sale's own financer (separate master from vehicle financers)
    financer_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("sale_financers.id", ondelete="SET NULL")
    )
    financer: Mapped["SaleFinancer | None"] = relationship()  # noqa: F821

    @property
    def financer_name(self) -> str | None:
        """The sale financer's name, for display."""
        return self.financer.name if self.financer else None
